contracts/carrot-app/src/helpers.rs

```diff
@@ -1,23 +1,96 @@
+use std::collections::HashMap;
+
+use crate::contract::{App, AppResult};
+use crate::error::AppError;
+use crate::exchange_rate::query_exchange_rate;
+use abstract_app::traits::AccountIdentification;
 use abstract_app::{objects::AssetEntry, traits::AbstractNameService};
 use abstract_sdk::Resolve;
-use cosmwasm_std::{Addr, Deps, Uint128};
-
-use crate::{
-    contract::{App, AppResult},
-    error::AppError,
-};
+use cosmwasm_std::{Addr, Coin, Coins, Decimal, Deps, Env, MessageInfo, StdResult, Uint128};
 
-pub fn get_user(deps: Deps, app: &App) -> AppResult<Addr> {
-    Ok(app
-        .admin
-        .query_account_owner(deps)?
-        .admin
-        .ok_or(AppError::NoTopLevelAccount {})
-        .map(|admin| deps.api.addr_validate(&admin))??)
+pub fn assert_contract(info: &MessageInfo, env: &Env) -> AppResult<()> {
+    if info.sender == env.contract.address {
+        Ok(())
+    } else {
+        Err(AppError::Unauthorized {})
+    }
 }
 
 pub fn get_balance(a: AssetEntry, deps: Deps, address: Addr, app: &App) -> AppResult<Uint128> {
     let denom = a.resolve(&deps.querier, &app.ans_host(deps)?)?;
     let user_gas_balance = denom.query_balance(&deps.querier, address.clone())?;
     Ok(user_gas_balance)
 }
+
+pub fn get_proxy_balance(deps: Deps, app: &App, denom: String) -> AppResult<Coin> {
+    Ok(deps
+        .querier
+        .query_balance(app.account_base(deps)?.proxy, denom.clone())?)
+}
+
+pub fn add_funds(funds: Vec<Coin>, to_add: Coin) -> StdResult<Vec<Coin>> {
+    let mut funds: Coins = funds.try_into()?;
+    funds.add(to_add)?;
+    Ok(funds.into())
+}
+
+pub const CLOSE_PER_MILLE: u64 = 1;
+
+/// Returns wether actual is close to expected within CLOSE_PER_MILLE per mille
+pub fn close_to(expected: Decimal, actual: Decimal) -> bool {
+    let close_coeff = Decimal::permille(CLOSE_PER_MILLE);
+
+    if expected == Decimal::zero() {
+        return actual < close_coeff;
+    }
+
+    actual > expected * (Decimal::one() - close_coeff)
+        && actual < expected * (Decimal::one() + close_coeff)
+}
```

Comment:
```suggestion
pub const CLOSE_PER_MILLE: Decimal = Decimal::permille(1);

/// Returns wether actual is close to expected within CLOSE_PER_MILLE per mille
pub fn close_to(expected: Decimal, actual: Decimal) -> bool {
    expected.abs_diff(actual) <= CLOSE_PER_MILLE
}
```
You can use permille const fn

Reply:
Yes thanks